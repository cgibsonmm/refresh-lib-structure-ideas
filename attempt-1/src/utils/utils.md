# Utils

This folder will house global used Utils

I.E.

1. commaFormat
