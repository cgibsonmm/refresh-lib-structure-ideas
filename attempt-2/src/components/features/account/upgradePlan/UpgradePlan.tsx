import React from 'react';
import { Box, Text } from '@/theme/components';
import { LimitReached } from './LimitReached/LimitReached';
import { PlanCard } from './PlanCard/PlanCard';
import {
  CustomSolutionButton,
  CustomSolutionContainer,
  CustomSolutionDescription,
  CustomSolutionHeading,
  Heading,
  PlansContainer,
  SubHeading,
  SubscriptionDescription,
} from './components';

export interface CurrentPlan {
  /** Plan name to show to the user. */
  title: string;
  /** Plan monthly limit. */
  limit: number | 'unlimited';
  /**
   * Monthly cost per report for the current plan. Use the `costPerReport` util to get this value since there are
   * important rounding considerations to generate the correct value.
   */
  costPerReport?: number;
  /** Indicates if the subscription is canceled. */
  canceled?: boolean;
}

export interface Plan {
  /** Plan unique key */
  key: string;
  title: string;
  description: string;
  /** Plan monthly cost */
  amount: number;
  /** Plan monthly limit, -1 for unlimited plans  */
  limit: number;
  /** Renewal period expressed in months */
  renewalPeriod: number;
  /** Monthly cost per report. Use the costPerReport util to calculate this. */
  costPerReport?: number;
  /** Indicates if this is the most popular plan */
  mostPopular?: boolean;
  /** Savings percentage vs current subscription. Use the savings util to calculate this. */
  savings?: number;
}

export interface UpgradePlanProps {
  /** Information about the current plan. */
  currentPlan: CurrentPlan;
  /** List of plans available for the user to upgrade to. */
  plans: Plan[];
  /** Callback fired when the user selects a plan from the list. */
  onSelectPlan: (plan: Plan) => void;
  /** Callback fired when the user clicks the Contact Us button. */
  onContactUs: () => void;
}

/**
 * This component is responsible for providing users with the ability to upgrade their plan.
 * It displays basic current plan information and also a list of plans available for the user to upgrade to.
 * It also provides ways for the user to **Contact Us** if they need a customized solution and show the user
 * once they have reached their limit when they are in the highest plan available.
 */
export function UpgradePlan({
  currentPlan: { title, limit, costPerReport = 0, canceled = false },
  plans,
  onSelectPlan,
  onContactUs,
}: UpgradePlanProps) {
  const noPlans = plans.length === 0;

  if (noPlans) {
    return <LimitReached onContactUs={onContactUs} />;
  }

  const subscriptionCostPerReport: string | JSX.Element =
    limit !== 'unlimited' && costPerReport > 0
      ? `($${costPerReport.toFixed(2)} per report)`
      : '';

  let subscriptionDescription: string | JSX.Element;

  if (canceled) {
    subscriptionDescription = 'Subscribe to run more reports today';
  } else {
    subscriptionDescription = (
      <>
        Your current plan is{' '}
        <Text
          component="span"
          sx={(theme) => ({
            color: theme.palette.primary.main,
            fontSize: 'inherit',
          })}
        >
          {title}
        </Text>{' '}
        and includes{' '}
        <Text
          component="span"
          sx={(theme) => ({
            color: theme.palette.primary.main,
            fontSize: 'inherit',
          })}
        >
          {limit} report credits {subscriptionCostPerReport}
        </Text>
      </>
    );
  }

  return (
    <Box>
      <Box sx={{ paddingInline: 'clamp(20px, 10vw, 150px)' }}>
        <Heading component="h1" variant="h1">
          Unlock more reports and lower your cost per report
        </Heading>
        <SubHeading variant="h2">
          Select a new plan and expand your search capabilities
        </SubHeading>
        <SubscriptionDescription>
          {subscriptionDescription}
        </SubscriptionDescription>
      </Box>

      <PlansContainer>
        {plans.map((plan) => (
          <PlanCard
            key={plan.key}
            heading={plan.title}
            subheading={plan.description}
            limit={plan.limit}
            cost={plan.amount}
            renewalPeriod={plan.renewalPeriod}
            costPerReport={plan.costPerReport}
            mostPopular={plan.mostPopular}
            savings={plan.savings}
            hoverCard={true}
            onClick={() => onSelectPlan(plan)}
          />
        ))}
      </PlansContainer>

      <CustomSolutionContainer>
        <Box>
          <CustomSolutionHeading>
            Need a customized solution?
          </CustomSolutionHeading>
          <CustomSolutionDescription>
            Still want more?
          </CustomSolutionDescription>
          <CustomSolutionDescription>
            Get a customized plan for your needs!
          </CustomSolutionDescription>
        </Box>
        <CustomSolutionButton
          variant="contained"
          color="success"
          size="large"
          onClick={onContactUs}
        >
          Contact us now
        </CustomSolutionButton>
      </CustomSolutionContainer>
    </Box>
  );
}
