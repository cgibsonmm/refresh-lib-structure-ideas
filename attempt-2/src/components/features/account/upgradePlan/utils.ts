/**
 * Calculates the cost per month for a plan
 *
 * @param cost Cost of the plan
 * @param renewalPeriod Number of months the plan is valid for
 * @returns Cost per month
 */
export function monthlyPlanCost(cost: number, renewalPeriod: number): number {
  return cost / renewalPeriod;
}

/**
 * Calculates the monthly cost per report for a plan
 *
 * @param cost Cost of the plan
 * @param renewalPeriod Number of months the plan is valid for
 * @param limit Monthly report limit of the plan
 * @returns Cost per report rounded to the nearest hundredth
 */
export function costPerReport(
  cost: number,
  renewalPeriod: number,
  limit: number
): number {
  const mthlyPlanCost = Math.ceil(monthlyPlanCost(cost, renewalPeriod));

  // Do not change, the rounding used here is important and was agreed with the product team
  //round to hundredths
  const costPerReport = Math.ceil((mthlyPlanCost / limit) * 100) / 100;

  return costPerReport;
}

/**
 * Calculates the savings based on the monthly cost per report of upgrading from a base plan to a new plan
 *
 * @param baseCost Cost of the base plan
 * @param baseRenewalPeriod Number of months the base plan is valid for
 * @param baseLimit Monthly report limit of the base plan
 * @param upgradeCost Cost of the upgrade plan
 * @param upgradeRenewalPeriod Number of months the new plan is valid for
 * @param upgradeLimit Monthly report limit of the upgrade plan
 * @returns Savings as a percentage
 */
export function savings(
  baseCost: number,
  baseRenewalPeriod: number,
  baseLimit: number,
  upgradeCost: number,
  upgradeRenewalPeriod: number,
  upgradeLimit: number
): number {
  const baseCostPerReport = costPerReport(
    baseCost,
    baseRenewalPeriod,
    baseLimit
  );
  const upgradeCostPerReport = costPerReport(
    upgradeCost,
    upgradeRenewalPeriod,
    upgradeLimit
  );

  // Do not change, the rounding used here is important and was agreed with the product team
  const savings = Math.floor(
    ((baseCostPerReport - upgradeCostPerReport) / baseCostPerReport) * 100
  );

  return savings;
}
