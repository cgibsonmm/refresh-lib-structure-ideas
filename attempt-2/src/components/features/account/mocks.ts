import { rest } from 'msw';
import { mockResolvers } from '@/utils/TestUtils';
import { fetchAccountMock } from './account.mock';
import {
  RESET_PASSWORD_ENDPOINT,
  REQUEST_RESET_PASSWORD_EMAIL_ENDPOINT,
} from './constants';
import { FETCH_ACCOUNT_ENDPOINT } from './useAccount';
import { UPDATE_SETTING_ENDPOINT } from './useUserSetting';
import type { RestRequest, RestContext, ResponseResolver } from 'msw';

const requestResetPasswordEmailResponses = {
  success: {
    account: {
      message:
        'Great. We will send you instructions if an account with that email exists.',
    },
    meta: { status: 200 },
  },
  badRequest: {
    account: { errors: ['Missing Parameters'] },
    meta: { status: 400 },
  },
};

const fetchAccountResponses = {
  success: fetchAccountMock,
  badRequest: {
    account: { errors: ['Missing Parameters'] },
    meta: { status: 400 },
  },
};

const updateSettingResponses = {
  success: {
    message: 'Ok',
    meta: { status: 200 },
  },
  badRequest: {
    message: 'Bad Request',
    meta: { status: 400 },
  },
};

const resetPasswordResponses = {
  success: {},
  badRequest: {
    account: {
      messages: [
        'The reset token has expired. Please try to reset your password again.',
      ],
    },
    meta: { status: 400 },
  },
};

const sendCancellationRequestResponses = {
  success: {},
  badRequest: {
    account: {
      messages: ['Could not cancel account'],
    },
  },
};

const requestResetPasswordEmailResolvers = mockResolvers(
  requestResetPasswordEmailResponses,
  {
    default: async (req, res, ctx) => {
      const body = await req.arrayBuffer();
      const parsedBody = new URLSearchParams(body.toString());

      const isInvalidContentType =
        req.headers.get('Content-Type') !== 'application/x-www-form-urlencoded';
      const isMissingBody = !parsedBody;
      // Because of how we're parsing the request body here, undefined email values were being converted to the string "undefined," and failing to trigger the error response. Adding an extra check for the literal string "undefined" solves the issue, but this may be an indication that we should be parsing 'x-www-form-urlencoded' requests differently in our mocks. - c.j.
      const isMissingEmail =
        parsedBody.get('account[email]') === (undefined || 'undefined');
      const isInvalidRequest =
        isInvalidContentType || isMissingBody || isMissingEmail;

      if (isInvalidRequest)
        return res(
          ctx.status(400),
          ctx.json(requestResetPasswordEmailResponses.badRequest)
        );
      else
        return res(
          ctx.status(200),
          ctx.json(requestResetPasswordEmailResponses.success)
        );
    },
  }
);

const resetPasswordResolvers = mockResolvers(resetPasswordResponses, {
  default: async (req, res, ctx) => {
    const validPassword = 'VALID_TEST_PASSWORD';
    const validToken = 'VALID_TEST_TOKEN';
    const body = await req.arrayBuffer();
    const parsedBody = new URLSearchParams(body.toString());
    const isInvalidContentType =
      req.headers.get('Content-Type') !== 'application/x-www-form-urlencoded';
    const hasInvalidParams =
      parsedBody.get('account[password]') !== validPassword ||
      parsedBody.get('account[password_confirmation]') !== validPassword ||
      parsedBody.get('password_reset_token') !== validToken;
    const isInvalidRequest = isInvalidContentType || hasInvalidParams;
    if (isInvalidRequest)
      return res(ctx.status(400), ctx.json(resetPasswordResponses.badRequest));
    else return res(ctx.status(200), ctx.json(resetPasswordResponses.success));
  },
});

const sendCancellationRequestResolvers = mockResolvers(
  sendCancellationRequestResponses,
  {
    default: async (req, res, ctx) => {
      return res(
        ctx.status(200),
        ctx.json(sendCancellationRequestResponses.success)
      );
    },
  }
);

const fetchAccountResolvers = mockResolvers(fetchAccountResponses);

type MockResolver = ResponseResolver<RestRequest<URLSearchParams>, RestContext>;

function requestResetPasswordEmailHandler(resolver: MockResolver) {
  return rest.post(REQUEST_RESET_PASSWORD_EMAIL_ENDPOINT, resolver);
}

function resetPasswordHandler(resolver: MockResolver) {
  return rest.post(RESET_PASSWORD_ENDPOINT, resolver);
}

function fetchAccountHandler(resolver: MockResolver) {
  const fetchAccountUrl = FETCH_ACCOUNT_ENDPOINT;
  return rest.get(fetchAccountUrl, resolver);
}

const updateSettingResolvers = mockResolvers(updateSettingResponses);

function updateSettingHandler(resolver: MockResolver) {
  const updateSettingUrl = UPDATE_SETTING_ENDPOINT;
  return rest.patch(updateSettingUrl, resolver);
}

function sendCancellationRequestHandler(resolver: MockResolver) {
  const sendCancellationRequestUrl = '/api/v5/account/cancel';
  return rest.post(sendCancellationRequestUrl, resolver);
}

export const handlers = {
  requestResetPasswordEmail: requestResetPasswordEmailHandler,
  resetPassword: resetPasswordHandler,
  fetchAccount: fetchAccountHandler,
  updateSetting: updateSettingHandler,
  sendCancellationRequest: sendCancellationRequestHandler,
};

export const resolvers = {
  requestResetPasswordEmail: requestResetPasswordEmailResolvers,
  resetPassword: resetPasswordResolvers,
  fetchAccount: fetchAccountResolvers,
  updateSetting: updateSettingResolvers,
  sendCancellationRequest: sendCancellationRequestResolvers,
};

export const responses = {
  requestResetPasswordEmail: requestResetPasswordEmailResponses,
  resetPassword: resetPasswordResponses,
  fetchAccount: fetchAccountResponses,
  updateSetting: updateSettingResponses,
  sendCancellationRequest: sendCancellationRequestResponses,
};
