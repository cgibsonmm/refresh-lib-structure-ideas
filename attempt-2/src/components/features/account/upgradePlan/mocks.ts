import { Plan, CurrentPlan } from './UpgradePlan';
import { costPerReport, savings } from './utils';

export const basePlanMock: Plan = {
  key: '26_89_1_month_nofree_afill_freshness_limit_100_freepdf_upsell',
  title: 'Starter',
  description: 'Starting small and simple',
  amount: 26.89,
  limit: 100,
  renewalPeriod: 1,
};

export const plansMock: Plan[] = [
  {
    key: '36_58_1_month_nofree_afill_freshness_limit_200_freepdf_upsell',
    title: 'Standard',
    description: 'Double your search capacity',
    amount: 36.58,
    limit: 200,
    renewalPeriod: 1,
    costPerReport: costPerReport(36.58, 1, 200),
    savings: savings(
      basePlanMock.amount,
      basePlanMock.renewalPeriod,
      basePlanMock.limit,
      36.58,
      1,
      200
    ),
  },
  {
    key: '63_99_1_month_nofree_afill_freshness_limit_400_freepdf_upsell',
    title: 'Premium',
    description: 'Your premium reports package',
    amount: 63.99,
    limit: 400,
    renewalPeriod: 1,
    costPerReport: costPerReport(63.99, 1, 400),
    mostPopular: true,
    savings: savings(
      basePlanMock.amount,
      basePlanMock.renewalPeriod,
      basePlanMock.limit,
      63.99,
      1,
      400
    ),
  },
  {
    key: '114_29_1_month_nofree_afill_freshness_limit_1000_freepdf_upsell',
    title: 'Professional',
    description: 'Take your business to the next level',
    amount: 114.29,
    limit: 1000,
    renewalPeriod: 1,
    costPerReport: costPerReport(114.29, 1, 1000),
    savings: savings(
      basePlanMock.amount,
      basePlanMock.renewalPeriod,
      basePlanMock.limit,
      114.29,
      1,
      1000
    ),
  },
];

export const currentPlanMock: CurrentPlan = {
  title: basePlanMock.title,
  limit: basePlanMock.limit,
  costPerReport: costPerReport(
    basePlanMock.amount,
    basePlanMock.renewalPeriod,
    basePlanMock.limit
  ),
};
