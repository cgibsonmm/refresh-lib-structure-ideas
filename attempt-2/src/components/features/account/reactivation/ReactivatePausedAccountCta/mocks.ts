import { fetchAccountMock } from '@/account/account.mock';

export const pausedAccountMock = {
  ...fetchAccountMock,
  account: {
    ...fetchAccountMock.account,
    subscription_info: {
      ...fetchAccountMock.account.subscription_info,
      monthly_reports_remaining: 100,
      monthly_report_limit: 100,
    },
  },
  meta: {
    ...fetchAccountMock.meta,
    subscription_state: 'paused',
  },
};
