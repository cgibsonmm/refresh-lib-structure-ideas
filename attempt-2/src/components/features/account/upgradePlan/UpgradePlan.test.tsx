import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { disableReactQueryErrorLogs } from '@/utils/TestUtils';
import { UpgradePlan, UpgradePlanProps } from './UpgradePlan';
import { currentPlanMock, plansMock } from './mocks';

disableReactQueryErrorLogs();

const props: UpgradePlanProps = {
  currentPlan: currentPlanMock,
  plans: plansMock,
  onSelectPlan: jest.fn(),
  onContactUs: jest.fn(),
};

test('renders the correct content', () => {
  const { asFragment } = render(<UpgradePlan {...props} />);

  expect(asFragment()).toMatchSnapshot();
});

test('renders the limit reached view', () => {
  const { container } = render(<UpgradePlan {...props} plans={[]} />);

  expect(container.getElementsByClassName('limit-reached').length).toBe(1);
});

test('renders current plan and upgrade options', () => {
  const { container } = render(<UpgradePlan {...props} />);

  expect(
    screen.getByRole('heading', {
      name: 'Unlock more reports and lower your cost per report',
      level: 1,
    })
  ).toBeInTheDocument();
  expect(
    screen.getByRole('heading', {
      name: 'Select a new plan and expand your search capabilities',
      level: 2,
    })
  ).toBeInTheDocument();

  expect(screen.getByText(/^Your current plan is/)).toBeInTheDocument();
  expect(screen.getByText('Starter')).toBeInTheDocument();
  expect(screen.getByText(/and includes/)).toBeInTheDocument();
  expect(
    screen.getByText('100 report credits ($0.27 per report)')
  ).toBeInTheDocument();

  expect(container.getElementsByClassName('plan-card').length).toBe(3);
  expect(screen.getByText('Need a customized solution?')).toBeInTheDocument();
  expect(screen.getByText('Still want more?')).toBeInTheDocument();
  expect(
    screen.getByText('Get a customized plan for your needs!')
  ).toBeInTheDocument();
  expect(
    screen.getByRole('button', { name: 'Contact us now' })
  ).toBeInTheDocument();
});

test('renders correct limit when subscription is unlimited', () => {
  render(
    <UpgradePlan
      {...props}
      currentPlan={{ ...currentPlanMock, limit: 'unlimited' }}
    />
  );

  expect(screen.getByText('unlimited report credits')).toBeInTheDocument();
});

test('renders cost per report only when provided', () => {
  render(
    <UpgradePlan
      {...props}
      currentPlan={{ ...currentPlanMock, costPerReport: undefined }}
    />
  );

  expect(screen.queryByText(/ per report\)$/)).not.toBeInTheDocument();
});

test('renders subscription description when canceled', () => {
  render(
    <UpgradePlan
      {...props}
      currentPlan={{ ...currentPlanMock, canceled: true }}
    />
  );

  expect(screen.queryByText(/^Your current plan is/)).not.toBeInTheDocument();
  expect(
    screen.getByText('Subscribe to run more reports today')
  ).toBeInTheDocument();
});

test('allows selecting a plan', async () => {
  render(<UpgradePlan {...props} />);

  const upgradeButtons = screen.getAllByRole('button', {
    name: 'Upgrade plan',
  });

  await userEvent.click(upgradeButtons[0]);

  expect(props.onSelectPlan).toHaveBeenCalledWith(plansMock[0]);
});

test('allows contacting us', async () => {
  render(<UpgradePlan {...props} />);

  await userEvent.click(screen.getByRole('button', { name: 'Contact us now' }));

  expect(props.onContactUs).toHaveBeenCalled();
});
