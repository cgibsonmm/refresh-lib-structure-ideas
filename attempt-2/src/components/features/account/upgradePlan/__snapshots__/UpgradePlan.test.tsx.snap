// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct content 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-1l14s7c"
    >
      <h1
        class="MuiTypography-root MuiTypography-h1 css-xlu8vr-MuiTypography-root"
      >
        Unlock more reports and lower your cost per report
      </h1>
      <h2
        class="MuiTypography-root MuiTypography-h2 css-14yvy03-MuiTypography-root"
      >
        Select a new plan and expand your search capabilities
      </h2>
      <p
        class="MuiTypography-root MuiTypography-body1 css-9kxc8q-MuiTypography-root"
      >
        Your current plan is 
        <span
          class="MuiTypography-root MuiTypography-body1 css-1r8nnji-MuiTypography-root"
        >
          Starter
        </span>
         and includes 
        <span
          class="MuiTypography-root MuiTypography-body1 css-1r8nnji-MuiTypography-root"
        >
          100 report credits ($0.27 per report)
        </span>
      </p>
    </div>
    <div
      class="MuiBox-root css-fmj022"
    >
      <div
        class="plan-card MuiBox-root css-1274504"
      >
        <div
          class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-filledDefault css-i6f8j9-MuiChip-root"
        >
          <span
            class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
          >
            29% Savings
          </span>
        </div>
        <div>
          <h2
            class="MuiTypography-root MuiTypography-h2 css-z9vhew-MuiTypography-root"
          >
            Standard
          </h2>
          <p
            class="MuiTypography-root MuiTypography-body2 css-lj764l-MuiTypography-root"
          >
            Double your search capacity
          </p>
        </div>
        <p
          class="MuiTypography-root MuiTypography-body1 css-d886p9-MuiTypography-root"
        >
          <span>
            200 reports
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 css-r1ig40-MuiTypography-root"
          >
            /mo
          </span>
        </p>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
        />
        <div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1gn6kd-MuiTypography-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
            >
              $36.58
            </span>
            <span
              class="MuiTypography-root MuiTypography-body1 css-mquo0y-MuiTypography-root"
            >
              /mo
            </span>
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-14nz9tk-MuiTypography-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
            >
              29% savings, 
            </span>
            <span>
              $0.19 per report
            </span>
          </p>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge css-mkazzp-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Upgrade plan
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="plan-card most-popular MuiBox-root css-1274504"
      >
        <div
          class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-filledDefault css-122ex86-MuiChip-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorInherit css-i4bv87-MuiSvgIcon-root"
            data-testid="StarIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
            />
          </svg>
          <span
            class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
          >
            Most popular
          </span>
        </div>
        <div
          class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-filledDefault css-i6f8j9-MuiChip-root"
        >
          <span
            class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
          >
            40% Savings
          </span>
        </div>
        <div>
          <h2
            class="MuiTypography-root MuiTypography-h2 css-z9vhew-MuiTypography-root"
          >
            Premium
          </h2>
          <p
            class="MuiTypography-root MuiTypography-body2 css-lj764l-MuiTypography-root"
          >
            Your premium reports package
          </p>
        </div>
        <p
          class="MuiTypography-root MuiTypography-body1 css-d886p9-MuiTypography-root"
        >
          <span>
            400 reports
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 css-r1ig40-MuiTypography-root"
          >
            /mo
          </span>
        </p>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
        />
        <div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1gn6kd-MuiTypography-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
            >
              $63.99
            </span>
            <span
              class="MuiTypography-root MuiTypography-body1 css-mquo0y-MuiTypography-root"
            >
              /mo
            </span>
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-14nz9tk-MuiTypography-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
            >
              40% savings, 
            </span>
            <span>
              $0.16 per report
            </span>
          </p>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge css-mkazzp-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Upgrade plan
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="plan-card MuiBox-root css-1274504"
      >
        <div
          class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-filledDefault css-i6f8j9-MuiChip-root"
        >
          <span
            class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
          >
            55% Savings
          </span>
        </div>
        <div>
          <h2
            class="MuiTypography-root MuiTypography-h2 css-z9vhew-MuiTypography-root"
          >
            Professional
          </h2>
          <p
            class="MuiTypography-root MuiTypography-body2 css-lj764l-MuiTypography-root"
          >
            Take your business to the next level
          </p>
        </div>
        <p
          class="MuiTypography-root MuiTypography-body1 css-d886p9-MuiTypography-root"
        >
          <span>
            1000 reports
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 css-r1ig40-MuiTypography-root"
          >
            /mo
          </span>
        </p>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
        />
        <div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1gn6kd-MuiTypography-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
            >
              $114.29
            </span>
            <span
              class="MuiTypography-root MuiTypography-body1 css-mquo0y-MuiTypography-root"
            >
              /mo
            </span>
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-14nz9tk-MuiTypography-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
            >
              55% savings, 
            </span>
            <span>
              $0.12 per report
            </span>
          </p>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge css-mkazzp-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Upgrade plan
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiBox-root css-1n7s66"
    >
      <div
        class="MuiBox-root css-0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1vnoz0t-MuiTypography-root"
        >
          Need a customized solution?
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-zf8vzv-MuiTypography-root"
        >
          Still want more?
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-zf8vzv-MuiTypography-root"
        >
          Get a customized plan for your needs!
        </p>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge css-1sbu2i7-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Contact us now
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
